Sources/NIOTLS/NIOTypedApplicationProtocolNegotiationHandler.swift

```diff
@@ -80,6 +80,10 @@ public final class NIOTypedApplicationProtocolNegotiationHandler<NegotiationResu
         }
     }
 
+    deinit {
+        self.negotiatedPromise.fail(ChannelError.inappropriateOperationForState)
+    }
```

Comment:
This solution seems wrong. We should rely on a channel lifecycle event to do this, probably `handlerRemoved`, and we should also know whether we already completed a promise or not.

Reply:
If we do the channel lifecycle we should also create the promise in handlerAdded but then we cannot really fulfill the protocol requirements and I didn’t want to make it optional in the protocol. That’s why I moved it to deinit. I also had it in handlerRemoved initially.

I agree on the second point.

Reply:
I'm fine having this auto-fail on deinit _if_ we guarantee we only complete this once.

Reply:
same here!